### PR TITLE
Flav/resource string ids

### PR DESCRIPTION
**Pull request by flvndvd**
## Description

See context [here](https://dust4ai.slack.com/archives/C050SM8NSPK/p1720008294140389) and [here](https://dust4ai.slack.com/archives/C050SM8NSPK/p1720030710397119).

This PR introduces a new way to generate string Id for resources. Currently, most tables already include a `sId` column. The main issue with the current algorithm is that it requires a database roundtrip to exchange an `sId` for a `ModelId`. This happens frequently as we use `sId` in various places outside our infrastructure, such as storing files in GCS using `sId`.

To avoid this issue, we propose an algorithm that encodes a `ModelId` into a string ID and decodes a string ID back into a `ModelId`. This PR implements this new algorithm. We want to leverage these string IDs more and include additional data. The new string ID represents the following:
```
[<shard_key>, <workspace_id>, <resource_id>]
```
The goal is to use these IDs as we scale.

Constraints for this algorithm:
- Security is not a concern. The primary goal is to expose a more user-friendly ID and embed extra data.
- We want a minimum length of 10 characters, similar to the current `sId`.

To achieve this without reinventing the wheel, we use https://sqids.org/, which generates short and unique identifiers. 

⚠️ Sqids guarantees that encoding different numbers will produce unique IDs. However, due to the algorithm's design, decoding random IDs can sometimes produce the same numbers. For example (fake data, just to illustrate):
```
[1, 2, 3] -> abc
abc -> [1, 2, 3]
def -> [1, 2, 3]
```
This is not an issue for us.

To illustrate how we plan to use this, this PR also updates the recently added `FileResource` to use this approach. Since it is not yet used in production, we can safely drop the legacy `sId` column in favor of public IDs.

Regarding the current `sId`, we will need to continue using it for existing links and references. However, we will start exposing public IDs for older resources. These can be easily differentiated as public IDs will have a resource prefix.

## Risk

Safe to rollback until we start using the file upload API.

## Deploy Plan

Once in production, run the migration to drop the `sId` column on the `files` table.
```
psql $FRONT_DATABASE_URI -f ./migrations/db/migration_35.sql
```

### CODE REVIEW

File: front/lib/api/files/preprocessing.ts [hunk 1]
```diff
@@ -38,7 +38,7 @@ const resizeAndUploadToFileStorage: PreprocessingFunction = async (
   } catch (err) {
     logger.error(
       {
-        fileId: file.sId,
+        fileId: file.id,
         workspaceId: auth.workspace()?.sId,
```
Comment (spolu) [inline front/lib/api/files/preprocessing.ts:42]: Let's call it fileModelId

File: front/lib/api/files/preprocessing.ts [hunk 2]
```diff
@@ -108,7 +108,7 @@ const extractTextFromPDF: PreprocessingFunction = async (
...
```
Comment (spolu) [inline front/lib/api/files/preprocessing.ts:112]: Same here

File: front/lib/resources/file_resource.ts [hunk 3]
```diff
@@ -197,14 +200,21 @@ export class FileResource extends BaseResource<FileModel> {
       .createReadStream();
   }
 
+  get publicId(): string {
+    return makePublicId("file", {
```
Comment (spolu) [inline front/lib/resources/file_resource.ts:204]: These should be called sId not publicId ?

File: front/lib/resources/public_ids.ts [hunk 4]
```diff
@@ -0,0 +1,86 @@
+import type { ModelId, Result } from "@dust-tt/types";
```
Comment (spolu) [inline front/lib/resources/public_ids.ts:1]: Let's not introduce a new terminology this is just a kind of sId no?

File: front/lib/resources/public_ids.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+const RESOURCES_PREFIX = {
+  file: "fil",
+};
+
+const ALL_RESOURCES_PREFIXES = Object.values(RESOURCES_PREFIX);
+
+type ResourceNameType = keyof typeof RESOURCES_PREFIX;
+
+export function makePublicId(
```
Comment (spolu) [inline front/lib/resources/public_ids.ts:24]: makeSId ? :-)

File: front/lib/resources/public_ids.ts [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  resourceName: ResourceNameType,
+  {
+    id,
+    workspaceId,
+  }: {
+    id: ModelId;
+    workspaceId: ModelId;
+  }
+): string {
+  const idsToEncode = [
```
Comment (spolu) [inline front/lib/resources/public_ids.ts:34]: Let's add a region here. 1 = US

File: front/lib/resources/storage/models/files.ts [hunk 7]
```diff
@@ -27,7 +27,6 @@ export class FileModel extends Model<
   declare fileName: string;
   declare fileSize: number;
   declare status: FileStatus;
-  declare sId: string;
```
Comment (spolu) [inline front/lib/resources/storage/models/files.ts:30]: I would suggest continue storing the sId we just know it's a decodable one for this resource

In the future all resources will use these sId but we will continue storing them

**Review by spolu — COMMENTED** [review]
Let's not introduce a new concept here and stick with sId vs "public id" and continue storing them.

It's just a new line of sId we will progressively transition to for all sIds

File: front/lib/resources/storage/models/files.ts (hunk 7 above)
Comment (flvndvd) [reply front/lib/resources/storage/models/files.ts:30]: Why storing them if you can re-compute them?

File: front/lib/resources/public_ids.ts (hunk 4 above)
Comment (flvndvd) [reply front/lib/resources/public_ids.ts:1]: I'd push for a new terminology here as this concept is different than the existing `sId` one. Having one work/terminology per concept is pretty efficient IMHO.

File: front/lib/resources/storage/models/files.ts (hunk 7 above)
Comment (flvndvd) [reply front/lib/resources/storage/models/files.ts:30]: Seen IRL.

File: front/lib/resources/string_ids.ts [hunk 8]
```diff
@@ -0,0 +1,99 @@
+import type { ModelId, Result } from "@dust-tt/types";
+import { Err, Ok } from "@dust-tt/types";
+import Sqids from "sqids";
+
+import logger from "@app/logger/logger";
```
Comment (spolu) [inline front/lib/resources/string_ids.ts:5]: can we move the legacy function here and rename it as legacy.

I would name this file sId.ts but not strongly held (just to limit vocabulary creep up)

File: front/lib/resources/file_resource.ts [hunk 9]
```diff
@@ -197,6 +197,13 @@ export class FileResource extends BaseResource<FileModel> {
       .createReadStream();
   }
```
Comment (spolu) [inline front/lib/resources/file_resource.ts:198]: We want to add a sIdToModelId here as well no?

**Review by spolu — COMMENTED** [review]
LGTM

Make sure to test this more than you would normally since this is a big big big trapdoor if it's broken :)

**Review by spolu — APPROVED** [review]

File: front/lib/resources/file_resource.ts (hunk 9 above)
Comment (flvndvd) [reply front/lib/resources/file_resource.ts:198]: Will add in a follow up PR when I'll start using it.

File: front/lib/resources/string_ids.ts (hunk 8 above)
Comment (flvndvd) [reply front/lib/resources/string_ids.ts:7]: Done

**Comment by flvndvd** [comment]
> Make sure to test this more than you would normally since this is a big big big trapdoor if it's broken :)

Thoroughly tested with the files 💪 !

**Pull request merged**